modules/samples/firealarm/component/ui/src/main/resources/jaggeryapps/devicemgt/app/units/cdmf.unit.device.type.firealarm.type-view/public/js/download.js

```diff
@@ -0,0 +1,195 @@
+                        },
+                        function (data) {
+                            doAction(data);
+                        }
+                );
+            } else if (deviceName) {
+                $('.controls').append('<label for="deviceName" generated="true" class="error" ' +
+                                      'style="display: inline-block;">Please enter at least 4 ' +
+                                      'characters.</label>');
+                $('.control-group').removeClass('success').addClass('error');
+            } else {
+                $('.controls').append('<label for="deviceName" generated="true" class="error" ' +
+                                      'style="display: inline-block;">This field is required.' +
+                                      '</label>');
+                $('.control-group').removeClass('success').addClass('error');
+            }
+        });
+
+        $("a#download-device-cancel-link").click(function () {
+            hidePopup();
+        });
+
+    });
+}
+
+function downloadAgent() {
+    var deviceName = "";
+    $('.new-device-name').each(function () {
+        if (this.value != "") {
+            deviceName = this.value;
+        }
+    });
+    var deviceType = "";
+    $('.deviceType').each(function () {
+        if (this.value != "") {
+            deviceType = this.value;
+        }
+    });
+    var sketchType = "";
+    $('.sketchType').each(function () {
+        if (this.value != "") {
+            sketchType = this.value;
+        }
+    });
+    var deviceNameFormat = /^[^~?!#$:;%^*`+={}\[\]\\()|<>,'"]{1,30}$/;
+    if (deviceName && deviceNameFormat.test(deviceName)) {
+        $(modalPopupContent).html($('#device-agent-downloading-content').html());
+        var successCallback = function (data) {
+            data = JSON.parse(data);
+            hidePopup();
+            window.location = "/devicemgt/api/devices/sketch/download/" + data.responseContent;
+        };
+        var generateLink = "/" + deviceType + "_mgt/manager/device/" + sketchType
+                           + "/generate_link?deviceName=" + deviceName;
+        invokerUtil.get(generateLink, successCallback, function (message) {
+            console.log(message.content);
+            hidePopup();
+            doAction(data);
+        });
+    } else {
+        $("#invalid-username-error-msg span").text("Invalid device name");
+        $("#invalid-username-error-msg").removeClass("hidden");
+    }
+}
+
+function doAction(data) {
+    //if it is saml redirection response
+    if (data.status == null) {
+        document.write(data);
+    }
+
+    if (data.status == "401") {
+        $(modalPopupContent).html($('#device-401-content').html());
+        $("#device-401-link").click(function () {
+            window.location = "/devicemgt/login";
+        });
+        showPopup();
+    } else if (data == "403") {
+        $(modalPopupContent).html($('#device-403-content').html());
```

Comment:
Be consistent on using quotes for string definitions. use double quotes as the standard.